provisioner/node_pools.go

```diff
@@ -123,13 +201,164 @@ func (p *AWSNodePoolProvisioner) Provision(ctx context.Context, nodePools []*api
+	if stdin != "" {
+		args = append(args, "-f", "-")
+	}
+
+	newApplyCommand := func() *exec.Cmd {
+		cmd := exec.Command(args[0], args[1:]...)
+		// prevent kubectl to find the in-cluster config
+		cmd.Env = []string{}
+		return cmd
+	}
+	var output string
+	applyManifest := func() error {
+		cmd := newApplyCommand()
+		if stdin != "" {
+			cmd.Stdin = strings.NewReader(stdin)
 		}
+		output, err = p.execManager.Run(ctx, p.logger, cmd)
+		return err
 	}
-	return true
+	err = backoff.Retry(applyManifest, backoff.WithMaxRetries(backoff.NewExponentialBackOff(), maxApplyRetries))
+	if err != nil {
+		return "", err
+	}
+	return output, nil
+}
+
+func (p *KarpenterNodePoolProvisioner) provisionNodePool(ctx context.Context, nodePool *api.NodePool, values map[string]interface{}) error {
+	template, err := p.generateNodePoolTemplate(nodePool, values, provisionersFileName)
+	if err != nil {
+		return err
+	}
+	if template == "" {
+		return nil
+	}
+	_, err = p.kubectlExecute(ctx, []string{"apply"}, template)
+	if err != nil {
+		return errors.Wrapf(err, "kubectl apply failed for node pool %s", nodePool.Name)
+	}
+	return nil
+}
+
+func (p *KarpenterNodePoolProvisioner) isKarpenterInstalled(ctx context.Context) (bool, error) {
+	_, dynamicClient, mapper, err := kubernetes.InitClients(p.cluster.APIServerURL, p.tokenSource)
+	if err != nil {
+		return false, err
+	}
+	gvr, err := kubernetes.ResolveKind(mapper, "CustomResourceDefinition")
+	if err != nil {
+		return false, err
+	}
+
+	_, err = dynamicClient.Resource(gvr).Get(ctx, "provisioners.karpenter.sh", metav1.GetOptions{})
+	if err != nil {
+		serr, ok := err.(*apiErrors.StatusError)
+		if ok && serr.Status().Code == http.StatusNotFound {
+			return false, nil
+		}
+		return false, err
+	}
+	return true, nil
+}
+
+func (p *KarpenterNodePoolProvisioner) Reconcile(ctx context.Context, updater updatestrategy.UpdateStrategy) error {
+	karpenterPools := p.cluster.KarpenterPools()
+	installed, err := p.isKarpenterInstalled(ctx)
+	if err != nil {
+		return err
+	}
+	if !installed {
+		// skip if karpenter is not installed
+		p.logger.Infof("skipping reconcilation of karpenter node pools as karpenter is not installed")
+		return nil
+	}
+
+	_, dynamicClient, mapper, err := kubernetes.InitClients(p.cluster.APIServerURL, p.tokenSource)
+	if err != nil {
+		return err
+	}
+	provisionerGvr, err := kubernetes.ResolveKind(mapper, "provisioner.karpenter.sh")
+	if err != nil {
+		return err
+	}
+	nodeTemplateGvr, err := kubernetes.ResolveKind(mapper, "awsnodetemplates.karpenter.k8s.aws")
+	if err != nil {
+		return err
+	}
+	existingProvisioners, err := dynamicClient.Resource(provisionerGvr).List(ctx, metav1.ListOptions{})
+	if err != nil {
+		return err
+	}
+
+	for _, p := range existingProvisioners.Items {
+		if !inNodePoolList(&api.NodePool{Name: p.GetName()}, karpenterPools) {
+			err := dynamicClient.Resource(provisionerGvr).Delete(ctx, p.GetName(), metav1.DeleteOptions{})
+			if err != nil {
+				return err
```

Comment:
If we fail here, we could end up just deleting the provisioner (and not the awstemplate). Do we care about that?

Reply:
What is the effect of having a leaked awstemplate resource?

Reply:
The most important resource is the provisioner. this is a synonym to node-pool in our setup. if the provisioner is not there the controller will delete all the nodes belonging to it and will stop creating more nodes

awstemplate is an abstraction to AWS launch template. it must be referenced in a provisioner to be used. so the impact is some bytes of storage in ETCD.

in addition, the CLM will fail and retry,  we can then see the problem and address the problematic resource to unblock CLM

Reply:
its low effort, so I did it.